### PR TITLE
Added new flowers on bonemeal use

### LINKED ISSUE #1074: Implement Flowers and biome-dependent grass bonemealing
http://minecraft.gamepedia.com/Flowers

### PR DESCRIPTION
**Pull request by 12xx12**
fixes #1074

### CODE REVIEW

File: src/Items/ItemDye.h [hunk 1]
```diff
@@ -216,51 +216,139 @@ class cItemDyeHandler :
...
-		- 20 % big grass (2-block tall grass)
-		- 60 % tall grass (1-block tall grass)
-		- 20 % flowers (biome dependent variants)
+		- 0 up to 8  big grass  (2-block tall grass)
+		- 8 up tp 24 tall grass (1-block tall grass)
+		- 0 up to 8  flowers    (biome dependent variants)
 	The new plants are spawned within 7 taxicab distance of a_BlockPos, on a grass block.
 	Broadcasts a particle for each new spawned plant. */
-	void growPlantsAround(cWorld & a_World, Vector3i a_BlockPos)
+	static void GrowPlantsAround(cWorld & a_World, Vector3i a_BlockPos)
```
Comment (tigerw) [inline src/Items/ItemDye.h:224]: I'd recommend a construction like:

```cpp
static std::optional<Vector3i> GetRandomPlantPosition(const Vector3i BlockPosition, const bool IsTallGrass)
{
  const int OffsetY = ...;
  if (!ValidHeight)
  {
    return {};
  }

  const Vector3i Offset(X, Y, Z);
  if (
    (GetBlock(BlockPosition + Offset) != GRASS) ||
    (....)
  )
  {
    return {};
  }

  return Offset;
}
```

looks neater than a lambda I think.

**Comment by tigerw** [comment]
seen, thanks. I am so behind on reviews this week...

**Comment by tigerw** [comment]
It seems like these changes spawn much fewer plants than before. This might be due to the code running fewer random position checks; before we were doing 40 guaranteed but now it's 3 random numbers added together with an average of 24 or something. I think we should rework this so it does a random walk (looking for grass blocks) from the clicked block, each time it needs to spawn a plant. This will be more like vanilla behaviour of spawning from the centre out, and up the chances of finding a suitable grass to spawn plants on, compared to picking random coordinates which tend to be in the air.

**Comment by 12xx12** [comment]
So we don't Grant all 40 plants always but we grant all the plants the random generator decided to grant in this bonemeal use?

**Comment by tigerw** [comment]
At least we try to. Unfortunately the wiki isn't very clear on this, we might have to invent something that looks reasonable.

**Comment by 12xx12** [comment]
I'm gonna do some thinking tomorrow and come back

**Comment by tigerw** [comment]
I made an attempt, could you check if it seems sane?

File: src/Items/ItemDye.h [hunk 2]
```diff
@@ -216,51 +216,185 @@ class cItemDyeHandler :
...
+			)
+		)
+		{
+			if (
+				!cChunkDef::IsValidHeight(Position.y) ||
+				(a_World.GetBlock(Position) != E_BLOCK_GRASS)  // Are we looking at grass?
+			)
+			{
+				// Not grass or invalid height, restart random walk and bail:
+				Position = a_Position;
```
Comment (12xx12) [inline src/Items/ItemDye.h:374]: Maybe also reset the the Tries variable so that we completely restart the tun

Comment (12xx12) [reply src/Items/ItemDye.h:374]: Although that might result in infinite loops when there's actually no space for a plant

**Review by tigerw — APPROVED** [review]

File: src/Items/ItemDye.h (hunk 2 above)
Comment (tigerw) [reply src/Items/ItemDye.h:374]: Yes, Tries acts a sort of a "fuel limit" for the loop as well as a walk distance limit

**Pull request merged**